### PR TITLE
feat(examples-plugins): add lighthouse to examples

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR includes:
- docs
- logic
- unit-tests
- integration-tests

**Question:**
How should we process multiple URLs?

closes [#448](https://github.com/code-pushup/cli/issues/448)

### CODE REVIEW

File: examples/plugins/src/lighthouse/src/constants.ts
Comment (vmasek) [inline examples/plugins/src/lighthouse/src/constants.ts:1]: I'd just put `/* eslint-disable no-magic-numbers */` on top of files that set these weight constatns and don't turn off the rule in general

File: code-pushup.config.ts [hunk 1]
```diff
@@ -56,6 +60,12 @@ const config: CoreConfig = {
       license: 'MIT',
       type: 'module',
     }),
+
+    lighthousePlugin({
+      url: 'https://quality-metrics-staging.web.app/login',
```
Comment (matejchalk) [inline code-pushup.config.ts:65]:
```suggestion
      url: 'https://staging.code-pushup.dev/login',
```

File: code-pushup.config.ts [hunk 2]
```diff
@@ -71,6 +81,11 @@ const config: CoreConfig = {
         { type: 'group', plugin: 'eslint', slug: 'suggestions', weight: 1 },
       ],
     },
+    {
+      slug: 'performance',
+      title: 'Performance',
+      refs: [...lighthouseCorePerfGroupRefs],
+    },
```
Comment (matejchalk) [inline code-pushup.config.ts:88]: Please put this under _Custom checks_ category, it's intended for these example plugins, so we that we can easily separate meaningful results for the CLI codebase (portal login page has nothing to do with CLI performance). I introduced this category in #399

File: examples/plugins/code-pushup.config.ts [hunk 3]
```diff
@@ -32,12 +36,22 @@ const config = {
         zod: '^3.22.4',
       },
     }),
+    lighthousePlugin({
+      url: 'https://quality-metrics-staging.web.app/login',
```
Comment (matejchalk) [inline examples/plugins/code-pushup.config.ts:40]:
```suggestion
      url: 'https://staging.code-pushup.dev/login',
```

File: examples/plugins/src/lighthouse/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# lighthouse-plugin
+
+🕵️ **Code PushUp plugin for lighthouse reports** 🔥
+
+---
+
+<img alt="Code PushUp plugin for lighthouse reports" src="./docs/images/lighthouse-plugin-cover.png" height="655">
+
+The plugin analyses a given URL and creates a lighthouse audits.
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:9]:
```suggestion
The plugin analyzes a given URL and creates Lighthouse audits.
```

File: examples/plugins/src/lighthouse/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# lighthouse-plugin
+
+🕵️ **Code PushUp plugin for lighthouse reports** 🔥
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:3]:
```suggestion
🕵️ **Code PushUp plugin for Lighthouse reports** 🔥
```

File: examples/plugins/src/lighthouse/README.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+- `url` - target to crawl
+- `outputPath` - path to lighthouse report in json format _(optional)_
+- `onlyAudits` - list of audits to run _(optional)_
+- `verbose` - additional information _(optional)_
+- `headless` - run headless _(optional)_
+
+## Getting started
+
+1. If you haven't already, install [@code-pushup/cli](../cli/README.md) and create a configuration file.
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:21]:
```suggestion
1. If you haven't already, install [@code-pushup/cli](../../../../packages/cli/README.md) and create a configuration file.
```

File: examples/plugins/src/lighthouse/README.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+2. Copy the [plugin source](../lighthouse) as is into your project
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:23]:
```suggestion
2. Copy the [plugin source](./src/) as is into your project
```

File: examples/plugins/src/lighthouse/README.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+3. Add this plugin to the `plugins` array in your Code PushUp CLI config file (e.g. `code-pushup.config.js`).
+
+   Pass in the path on the directory to crawl (relative to `process.cwd()`), as well as patterns and a budget.
+
+   ```js
+   import { LIGHTHOUSE_OUTPUT_FILE_DEFAULT } from './lighthouse-plugin.constants';
+   import lighthousePlugin from './lighthouse.plugin';
+
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:32]: If it's meant to be a working example:

```suggestion
import { join } from 'node:path';
```

File: examples/plugins/src/lighthouse/README.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+   import fileSizePlugin, { recommendedRefs as lighthouseRecommendedRefs } from './lighthouse.plugin';
+
+   export default {
+     // ...
+     categories: [
+       // ...
+       {
+         slug: 'performance',
+         title: 'Performance',
+         refs: [...lighthouseRecommendedRefs],
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:59]:
```suggestion
         refs: lighthouseRecommendedRefs,
```

File: examples/plugins/src/lighthouse/README.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+       },
+     ],
+   };
+   ```
+
+5. Run the CLI with `npx code-pushup collect` and view or upload report (refer to [CLI docs](../cli/README.md)).
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:65]:
```suggestion
5. Run the CLI with `npx code-pushup collect` and view or upload report (refer to [CLI docs](../../../../packages/cli/README.m)).
```

File: examples/plugins/src/lighthouse/README.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+## Audits
+
+Detailed information about the audits can be found in the docs of [lighthouse](@TODO).
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/README.md:69]: Missing link, could be this for example:

```suggestion
Detailed information about the audits can be found in the docs of [Lighthouse](https://developer.chrome.com/docs/lighthouse/overview/).
```

File: examples/plugins/src/lighthouse/mock/constants.ts [hunk 12]
```diff
@@ -0,0 +1 @@
+export const LIGHTHOUSE_URL = 'https://quality-metrics-staging.web.app/login';
```
Comment (matejchalk) [inline examples/plugins/src/lighthouse/mock/constants.ts:1]:
```suggestion
export const LIGHTHOUSE_URL = 'https://staging.code-pushup.dev/login';
```

File: examples/plugins/src/lighthouse/mock/chromium/Readme.md
Comment (matejchalk) [inline examples/plugins/src/lighthouse/mock/chromium/Readme.md:1]: What is this? :raised_eyebrow: It's seems to be some TS snippet, not Markdown.

File: packages/utils/src/lib/reports/scoring.ts [hunk 13]
```diff
@@ -50,6 +50,9 @@ export function scoreReport(report: Report): ScoredReport {
         `${slug}-${ref.slug}-audit`,
       )?.score;
       if (score == null) {
+        // @TODO error message is not well written
+        // suggestion:
+        // `Group has invalid ref - audit with slug ${ref.slug} from plugin ${plugin.slug} not found`,
         throw new Error(
           `Group has invalid ref - audit with slug ${slug}-${ref.slug}-audit not found`,
         );
```
Comment (matejchalk) [inline packages/utils/src/lib/reports/scoring.ts:58]: Go for it :+1:

File: .code-pushup/.gitkeep
Comment (matejchalk) [inline .code-pushup/.gitkeep:1]: Does this need to exist? I have PTSD from `tmp/.gitkeep` appearing in every diff :sweat_smile:

File: code-pushup.config.ts (hunk 2 above)
Comment (BioPhoton) [reply code-pushup.config.ts:88]: done

File: examples/plugins/src/lighthouse/mock/chromium/Readme.md
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/mock/chromium/Readme.md:1]: removed

File: packages/utils/src/lib/reports/scoring.ts (hunk 13 above)
Comment (BioPhoton) [reply packages/utils/src/lib/reports/scoring.ts]: done

File: packages/utils/src/lib/reports/__snapshots__/generate-stdout-summary.integration.test.ts.snap
Comment (matejchalk) [inline packages/utils/src/lib/reports/__snapshots__/generate-stdout-summary.integration.test.ts.snap:1]: Why was this changed? :raised_eyebrow:

Comment (matejchalk) [reply packages/utils/src/lib/reports/__snapshots__/generate-stdout-summary.integration.test.ts.snap:1]: @Tlacenka I guess it's related to #461?

Comment (Tlacenka) [reply packages/utils/src/lib/reports/__snapshots__/generate-stdout-summary.integration.test.ts.snap:1]: Indeed.

Comment (BioPhoton) [reply packages/utils/src/lib/reports/__snapshots__/generate-stdout-summary.integration.test.ts.snap:1]: merged main. Now the diff is gone.

**Review by matejchalk — APPROVED** [review]

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+      userDataDir: 'test',
+    });
+    expect(pluginConfig.runner.args).toEqual(
+      expect.arrayContaining([
+        '--chrome-flags="--headless=new --user-data-dir=test"',
+      ]),
+    );
+  });
+
+  it('should return PluginConfig that executes correctly', async () => {
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:102]: Tests of such a different scope (parsing arguments vs executing the whole plugin) should not be in the same file under the same `describe`.

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    const pluginConfig = await create({
+      url: LIGHTHOUSE_URL,
+      onlyAudits: 'largest-contentful-paint',
+    });
+    expect(pluginConfig.runner.args).toEqual(
+      expect.arrayContaining(['--onlyAudits="largest-contentful-paint"']),
+    );
+    const { audits: auditOutputs } = await executePlugin(pluginConfig);
+
+    expect(auditOutputs).toHaveLength(1);
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:127]: The exact audit slug `largest-contentful-paint` can be tested here, otherwise it can be any one audit and pass the test.

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+  });
+}, 30_000);
+
+describe('lighthouse-audits-export', () => {
+  it.each(audits)('should be a valid audit meta info', audit => {
+    expect(() => auditSchema.parse(audit)).not.toThrow();
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:133]: `%s` with audit slug should be added to the description to distinguish test cases here.

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+  });
+});
+
+describe('lighthouse-corePerfGroupRefs-export', () => {
+  it.each(corePerfGroupRefs)(
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:138]: Similarly to the audits, the group reference audit should be added to the test description, otherwise multiple test cases of the same name will be logged as output.

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    expect(runnerConfig(baseOptions)).toEqual(
+      expect.objectContaining({
+        args: expect.arrayContaining(['lighthouse', LIGHTHOUSE_URL]),
+        command: 'npx',
+        outputFile: LIGHTHOUSE_OUTPUT_FILE_DEFAULT,
+      }),
+    );
+  });
+
+  it('should execute if with output options', () => {
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts:28]: I don't understand the purpose of this test case. The only property in `baseOptions` variable is the URL, no output options are given as far as I can see.

File: examples/plugins/src/lighthouse/src/utils.unit.test.ts [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    const pluginConfig = await create({
+      url: LIGHTHOUSE_URL,
+      headless: true,
+    });
+    expect(pluginConfig.runner.args).toEqual(
+      expect.arrayContaining(['--chrome-flags="--headless=new"']),
+    );
+  });
+
+  it('should parse options for headless to new if false is given', async () => {
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/utils.unit.test.ts:92]: The `create` function with `headless` is already tested in `lighthouse.plugin.integration.test`. This is a duplicate.

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+  it('should parse options for headless by default to false', () => {
+    const args = runnerConfig({
+      url: LIGHTHOUSE_URL,
+    });
+    expect(args).toEqual(
+      expect.not.arrayContaining(['--chrome-flags="--headless=new"']),
+    );
+  });
+
+  it('should run headless "new" if set to true', () => {
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts:66]: This case is already tested in the integration test.

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    ).toEqual(
+      expect.objectContaining({
+        args: expect.arrayContaining([
+          `--onlyAudits="${lcpAuditOutputBase.slug}"`,
+        ]),
+      }),
+    );
+  });
+
+  it('should parse options for headless by default to false', () => {
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts:57]: This case is also already tested in the integration test.

File: examples/plugins/src/lighthouse/src/utils.unit.test.ts [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    const pluginConfig = await create({
+      url: LIGHTHOUSE_URL,
+      headless: false,
+    });
+    expect(pluginConfig.runner.args).toEqual(
+      expect.not.arrayContaining(['--chrome-flags="--headless=new"']),
+    );
+  });
+
+  it('should parse options for userDataDir correctly', async () => {
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/utils.unit.test.ts:102]: The `create` function with `userDataDir` already is tested in `lighthouse.plugin.integration.test`. This is a duplicate.

File: examples/plugins/src/lighthouse/src/utils.unit.test.ts [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+describe('getLighthouseCliArguments', () => {
+  it('should parse valid options', () => {
+    expect(
+      getLighthouseCliArguments({
+        url: 'https://code-pushup-portal.com',
+      }),
+    ).toEqual(expect.arrayContaining(['https://code-pushup-portal.com']));
+  });
+
+  it('should parse options for headless to new if true is given', async () => {
```
Comment (Tlacenka) [inline examples/plugins/src/lighthouse/src/utils.unit.test.ts:82]: The `create` function with `headless` is already tested in `lighthouse.plugin.integration.test`. This is a duplicate.

File: packages/utils/src/lib/reports/scoring.ts [hunk 24]
```diff
@@ -51,7 +51,7 @@ export function scoreReport(report: Report): ScoredReport {
       )?.score;
       if (score == null) {
         throw new Error(
-          `Group has invalid ref - audit with slug ${slug}-${ref.slug}-audit not found`,
+          `Group has invalid ref - audit with slug ${ref.slug} from plugin ${slug} not found`,
```
Comment (Tlacenka) [inline packages/utils/src/lib/reports/scoring.ts:54]: Note: Identical error message is also present in `perf` benchmarks. Not sure to which extent we want to maintain them though.

File: packages/utils/src/lib/verbose-utils.ts [hunk 25]
```diff
@@ -1,20 +1,20 @@
-function getLogVerbose(verbose: boolean) {
+function getLogVerbose(verbose = false) {
```
Comment (Tlacenka) [inline packages/utils/src/lib/verbose-utils.ts:1]: Now that the verbose functions have a default, there should be at least one test calling the function without an argument to cover this use case.

**Review by Tlacenka — COMMENTED** [review]
@BioPhoton 
The PR got merged as I was doing the review. Please address these points in the comments and if applicable, in a follow-up PR.

Ideally, unless someone else explicitly gives a review for tests, ping me or ask me to review them please.

File: packages/cli/README.md
Comment (Tlacenka) [inline packages/cli/README.md:1]: How does this PR close the #448 ?
I do not see any changes in this PR addressing it.

Comment (BioPhoton) [reply packages/cli/README.md:1]: I added a ensureDirectory call in the plugin

File: examples/plugins/src/lighthouse/src/utils.unit.test.ts (hunk 23 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/utils.unit.test.ts:82]: the options have a slightly different logic in the plugin config and the utils function. Outside the typing is `boolean` inside `false` or `new`

File: examples/plugins/src/lighthouse/README.md [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+         url: 'https://example.com',
+         outputPath: join('.code-pushup', LIGHTHOUSE_OUTPUT_FILE_DEFAULT),
+       }),
+     ],
+   };
+   ```
+
+4. (Optional) Reference audits (or groups) that you wish to include in custom categories (use `npx code-pushup print-config` to list audits and groups).
+
+   Assign weights based on what influence each audit and group should have on the overall category score (assign weight 0 to only include it for extra info, without influencing the category score).
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/README.md:48]: removed the inforamtion

File: examples/plugins/src/lighthouse/README.md [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+
+   ```js
+   import fileSizePlugin, { recommendedRefs as lighthouseRecommendedRefs } from './lighthouse.plugin';
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/README.md:51]: added export

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+    const pluginConfig = await create({
+      url: LIGHTHOUSE_URL,
+      headless: true,
+    });
+    expect(pluginConfig.runner.args).toEqual(
+      expect.arrayContaining(['--chrome-flags="--headless=new"']),
+    );
+  });
+
+  it('should parse options for headless to new if false is given', async () => {
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:80]: added " in runner args" to the statement

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts (hunk 21 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts:57]: It tests the runnerConfig function. IMHO correct.

File: examples/plugins/src/lighthouse/src/utils.unit.test.ts (hunk 22 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/utils.unit.test.ts:102]: good catch!! I test `getLighthouseCliArguments` now. :)

File: packages/utils/src/lib/verbose-utils.ts (hunk 25 above)
Comment (BioPhoton) [reply packages/utils/src/lib/verbose-utils.ts:1]: added test

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts (hunk 14 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:102]: done

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts (hunk 15 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:127]: done

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts (hunk 16 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:133]: done

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts (hunk 17 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/lighthouse.plugin.integration.test.ts:138]: done

File: examples/plugins/src/lighthouse/src/utils.unit.test.ts (hunk 19 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/utils.unit.test.ts:92]: fixed

File: examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts (hunk 20 above)
Comment (BioPhoton) [reply examples/plugins/src/lighthouse/src/lighthouse.plugin.unit.test.ts:66]: fixed
